### PR TITLE
Aggregate edge

### PR DESCRIPTION
**Pull request by francoishalbach**

### CODE REVIEW

File: hammers/query.py [hunk 1]
```diff
@@ -523,6 +523,33 @@ def blazar_set_non_reservable(db, node):
...
+    SELECT ca.id, l.id AS lid, ch.hypervisor_hostname
+    FROM blazar.computehost_allocations ca
+    JOIN blazar.reservations r
+    ON ca.reservation_id=r.id
+    JOIN blazar.leases l
+    ON r.lease_id=l.id
+    JOIN blazar.computehosts ch
+    ON ca.compute_host_id=ch.id
+    WHERE ca.deleted is Null
+       AND l.end_date < curdate() - interval 7 day
```
Comment (diurnalist) [inline hammers/query.py:539]: Why the extra buffer of one week? I'm not opposed to it, just wondering the reasoning

File: hammers/scripts/clean_old_aggregates.py [hunk 2]
```diff
@@ -6,12 +6,16 @@
 import re
 import requests
 import traceback
+import dateutil.parser
+#from datetime import datetime
```
Comment (diurnalist) [inline hammers/scripts/clean_old_aggregates.py:10]: Please remove dead code like this

File: hammers/scripts/clean_old_aggregates.py [hunk 3]
```diff
@@ -20,15 +24,25 @@
...
+    'port': '3306',
+    })
+mysqlargs.inject(parser)
 args = parser.parse_args(sys.argv[1:])
 auth = osapi.Auth.from_env_or_args(args=args)
+mysqlargs.extract(args)
+conn = mysqlargs.connect()
 
-# Get aggregates and leases
+#Get aggregates, leases, host_allocations
```
Comment (diurnalist) [inline hammers/scripts/clean_old_aggregates.py:39]: I'm not even sure this comment needs to exist, we could remove it. The code explains what it is doing well enough in my view

File: hammers/scripts/clean_old_aggregates.py [hunk 4]
```diff
@@ -63,21 +77,75 @@ def clear_aggregates(agg_list):
 
     return errors, report
 
+def orphan_find(allaggs):
+
+    # Find all hosts currently in aggregates
```
Comment (diurnalist) [inline hammers/scripts/clean_old_aggregates.py:82]: A much better example of a comment :)

File: hammers/scripts/clean_old_aggregates.py [hunk 5]
```diff
@@ -63,21 +77,75 @@ def clear_aggregates(agg_list):
...
+            hosts_from_aggs.append(host)
+
+    # Make list of ironic hosts not in any aggregate
+    ironic_hosts = osrest.ironic.nodes(auth, details=False).keys()
+    orphans = []
+    for host in ironic_hosts:
+        if host not in hosts_from_aggs:
+            host_id = [h['id'] for h in osrest.blazar.hosts(auth).values() if h['uid'] == host][0]
+            orphans.append(host_id)
+    return(orphans)
```
Comment (diurnalist) [inline hammers/scripts/clean_old_aggregates.py:95]: This style isn't used/recommended 

```suggestion
    return orphans
```

File: hammers/scripts/clean_old_aggregates.py [hunk 6]
```diff
@@ -63,21 +77,75 @@ def clear_aggregates(agg_list):
...
+
+def has_active_allocation(orph):
+
+    matching_allocs = [alloc for alloc in host_allocs if alloc['resource_id'] == orph]
+    if not matching_allocs:
+        return False
+    res = matching_allocs[0]['reservations'][0]['id']
+    return(res)
```
Comment (diurnalist) [inline hammers/scripts/clean_old_aggregates.py:103]:
```suggestion
    return res
```

File: hammers/scripts/clean_old_aggregates.py [hunk 7]
```diff
@@ -63,21 +77,75 @@ def clear_aggregates(agg_list):
...
+
+def del_expired_alloc(db, old_alloc):
+
+    ha_id = old_alloc['id']
+    l_id = old_alloc['lid']
+    hh = old_alloc['hypervisor_hostname']
+
+    msg=("Deleting host_allocation for host {} matching expired lease {}. ".format(hh,l_id) + "\n")
```
Comment (diurnalist) [inline hammers/scripts/clean_old_aggregates.py:111]: Returning the error message feels a bit weird, as it's hard to parse for any other function that may want to call this later. Suggestion: return a tuple

```python
return hh, l_id
```

Then you can create the message like this (note I changed the tense to past tense)

```python
for alloc in old_allocations:
    hostname, lease_id = del_expired_alloc(conn, alloc)
    reports.append((
        "Deleted host_allocation for host {} matching expired lease {}.".format(hostname, lease_id))

File: hammers/scripts/clean_old_aggregates.py [hunk 8]
```diff
@@ -63,21 +77,75 @@ def clear_aggregates(agg_list):
...
-        if report:
-            str_report = ''.join(report)
-
-            print(str_report)
-
+        errors, reports = clear_aggregates(aggregate_list)
+        orphan_list = orphan_find(aggregates)
+        reports = []
+        
+        if orphan_list:
```
Comment (diurnalist) [inline hammers/scripts/clean_old_aggregates.py:127]: Looks like `orphan_list` is always going to be a list. Since the logic doesn't do anything special when it's empty, I'd just skip this check and proceed to `for orphan in orphan_list`, which effectively won't execute if the list has 0 items.

File: hammers/scripts/clean_old_aggregates.py [hunk 9]
```diff
@@ -63,21 +77,75 @@ def clear_aggregates(agg_list):
...
+                else:
+                    destination_agg = [aggr['id'] for aggr in aggregates.values() if aggr['name'] == destiny][0]
+                    reports.append("Moving orphan host {} to destined aggregate {}.".format(orphan, destination_agg) + "\n")
+                    osrest.nova.aggregate_add_host(auth, destination_agg, host['hypervisor_hostname'])
+        
+
+        old_allocations = query.blazar_find_old_host_alloc(conn)
+        for alloc in old_allocations:
+            reports.append(del_expired_alloc(conn, alloc))
+        conn.db.commit()
```
Comment (diurnalist) [inline hammers/scripts/clean_old_aggregates.py:145]: Is all of this taking place within a transaction? I'm not sure that makes sense, if so -- we are changing state in other systems like Nova and it feels like each step should apply before proceeding to the next step. Otherwise the transaction gets rolled back and we've updated Nova, but the DB has been reverted.

File: hammers/scripts/clean_old_aggregates.py [hunk 10]
```diff
@@ -63,21 +77,75 @@ def clear_aggregates(agg_list):
...
+
+        if reports:
+            str_report = ''.join(reports)
```
Comment (diurnalist) [inline hammers/scripts/clean_old_aggregates.py:148]: One trick I do is 
```suggestion
            str_report = '\n'.join(reports)
```
... which turns each line into its own line.

File: hammers/scripts/clean_old_aggregates.py [hunk 11]
```diff
@@ -63,21 +77,75 @@ def clear_aggregates(agg_list):
...
+        errors, reports = clear_aggregates(aggregate_list)
+        orphan_list = orphan_find(aggregates)
+        reports = []
+        
+        if orphan_list:
+            for orphan in orphan_list:
+                destiny = has_active_allocation(orphan)
+                host = osrest.blazar.host(auth, orphan)
+                if destiny is None:
+                    reports.append("Error identifying allocation for orphan host {}.".format(orphan) + "\n")
```
Comment (diurnalist) [inline hammers/scripts/clean_old_aggregates.py:132]: Don't need the newlines here, see comment below :)

File: hammers/query.py (hunk 1 above)
Comment (francoishalbach) [reply hammers/query.py]: I thought in case we want to recover a lease but that doesn't really make sense, I'll remove it.
